File: benchmarks/tsb/bench_sample_weighted.ts
/**
 * Benchmark: sampleSeries with weights — weighted random sampling from a
 * 100k-element Series. Extends bench_sample_fn which tests unweighted sampling.
 * Outputs JSON: {"function": "sample_weighted", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, sampleSeries } from "../../src/index.ts";

const SIZE = 100_000;
const N_SAMPLE = 1_000;
const WARMUP = 5;
const ITERATIONS = 30;

const data = Array.from({ length: SIZE }, (_, i) => i * 0.5);
// Weights: higher values get more weight (triangular distribution)
const weights = Array.from({ length: SIZE }, (_, i) => (i + 1) / SIZE);

const s = new Series({ data });

for (let i = 0; i < WARMUP; i++) {
  sampleSeries(s, { n: N_SAMPLE, weights });
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  sampleSeries(s, { n: N_SAMPLE, weights });
  times.push(performance.now() - t0);
}

const total = times.reduce((a, b) => a + b, 0);
console.log(
  JSON.stringify({
    function: "sample_weighted",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_sample_weighted.py
"""
Benchmark: pandas Series.sample() with weights — weighted random sampling.
Mirrors tsb's sampleSeries with weights option.
Outputs JSON: {"function": "sample_weighted", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import pandas as pd

SIZE = 100_000
N_SAMPLE = 1_000
WARMUP = 5
ITERATIONS = 30

data = [i * 0.5 for i in range(SIZE)]
# Weights: higher values get more weight (triangular distribution)
weights = [(i + 1) / SIZE for i in range(SIZE)]

s = pd.Series(data)

for _ in range(WARMUP):
    s.sample(n=N_SAMPLE, weights=weights, replace=False)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    s.sample(n=N_SAMPLE, weights=weights, replace=False)
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({
    "function": "sample_weighted",
    "mean_ms": total_ms / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total_ms,
}))
